indexer/search_string_utils.cpp

```diff
@@ -94,11 +96,67 @@ char const * kStreetTokensSeparator = "\t -,.";
 
 /// @todo Move prefixes, suffixes into separate file (autogenerated).
 /// It's better to distinguish synonyms comparison according to language/region.
-
 class StreetsSynonymsHolder
 {
 public:
-  using TStrings = search::StringSet<UniChar, 8>;
+  struct BooleanSum
+  {
+    void Add(bool value) { m_value = m_value || value; }
+
+    template <typename ToDo>
+    void ForEach(ToDo && toDo) const
+    {
+      if (m_value)
+        toDo(m_value);
+    }
+
+    void Clear() { m_value = false; }
+
+    bool m_value = false;
+  };
+
+  template <typename Char, typename SubTree>
+  class Moves
+  {
+  public:
+    template <typename ToDo>
+    void ForEach(ToDo && toDo) const
+    {
+      for (auto const & subtree : m_subtrees)
+        toDo(subtree.first, *subtree.second);
+    }
+
+    SubTree * GetSubTree(Char const & c) const
+    {
+      for (auto const & subtree : m_subtrees)
+      {
+        if (subtree.first == c)
+          return subtree.second.get();
+      }
+      return nullptr;
+    }
+
+    SubTree & GetOrCreateSubTree(Char const & c, bool & created)
+    {
+      for (size_t i = 0; i < m_subtrees.size(); ++i)
+      {
+        if (m_subtrees[i].first == c)
+        {
+          created = false;
+          return *m_subtrees[i].second;
+        }
+      }
+
+      created = true;
+      m_subtrees.emplace_back(c, make_unique<SubTree>());
+      return *m_subtrees.back().second;
+    }
+
+    void Clear() { m_subtrees.clear(); }
+
+  private:
+    buffer_vector<pair<Char, std::unique_ptr<SubTree>>, 8> m_subtrees;
```

Comment:
Yes, I need. Please, take a look at the old implementation of this class - it used buffer_vector in nodes, as the branching factor is quite low, therefore it's ok to pay for unused space.